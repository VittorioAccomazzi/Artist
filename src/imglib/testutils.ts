import { Canvas, createCanvas, loadImage } from 'canvas';
import {Image2D, ImageUint8, isImage} from './imagebase'
import Histogram from './histogram'
import CanvasUtils, {SeqCanvas} from './canvasUtils'
import {imageHash} from 'image-hash'
import * as fs from 'fs'
import * as path from 'path'
import TensorField from './tensorField';


export const tmpFolder = 'tmp'
export const testImage1 = 'src/imglib/testimages/test1.jpg'
export const testImage2 = 'src/imglib/testimages/test2.jpg'
export const testImages = [ testImage1, testImage2]

export async function hash( image : Canvas | Image2D ) : Promise<string> {
    let canvas =  isImage(image) ? toCanvas(image) : image
    let b = canvas.toBuffer('image/png');
    let name = `testImage-${canvas.width}x${canvas.height}.png`
    return new Promise((res, rej)=>{
        imageHash({
            ext: 'image/png',
            data: b,
            name: name
        }, 128, true, (err: Error, val : string )=>{
            if( err ) rej(err)
            res(val)
        })
    })
}

export async function* getCanvases() : AsyncGenerator<[Canvas, CanvasRenderingContext2D, string]>{
    for( let img of testImages) {
        let image = await loadImage(img)
        let canvas= createCanvas(image.width, image.height);
        let ctx = canvas.getContext('2d')
        ctx.drawImage(image, 0, 0)
        yield [canvas, ctx, path.basename(img,'.jpg')]
    }
}

export function toCanvas(image : Image2D) : Canvas {
    let max = image.maxValue()
    let min = image.minValue()
    let img = image

    if ( max > 255 || min <0 ){
        // no need to convert
        let range = max - min
        let scale = range > 0 ? 254/range : 1
        scale = Math.min(1, scale) // to not scale up.
        let offset = -scale * min
        img = image.convertTo('Uint8', scale, offset)
    }  
    let width = img.width
    let height= img.height
    let canvas= createCanvas(width, height)
    let ctx = canvas.getContext('2d')
    let data = ctx.getImageData(0,0, width, height)
    let pixels= img.imagePixels
    let ptr =0
    for(let y=0; y<height; y++){
        let yOffset = y*width
        for( let x=0; x<width; x++){
            let p = Math.floor(pixels[yOffset+x])
            data.data[ptr++] = p // Red
            data.data[ptr++] = p // Green
            data.data[ptr++] = p // blue
            data.data[ptr++] = 255 // alpha
        }
    }
    ctx.putImageData(data,0,0)
    return canvas
}

export async function dumpImage( image : Image2D, name? : string ){
    let nCanvas = toCanvas(image)
    dumpCanvas(nCanvas,name)
}

export async function dumpCanvas(canvas : Canvas, name? : string ) : Promise<void> {
    let fullname = path.resolve(tmpFolder, name ? name: (await hash(canvas)).substr(0,25))
    if( ! fs.existsSync(tmpFolder)){
        fs.mkdirSync(tmpFolder)
    }
    await saveCanvas(canvas, fullname)
}

export function toSeqCanvas( nCanvas : Canvas ) : SeqCanvas {
    // this is inefficient, but allows to test more code
    let hCanvas = toHTMLCanvas(nCanvas)
    return CanvasUtils.toSeq(hCanvas)
}

export function toHTMLCanvas(nCanvas : Canvas) : HTMLCanvasElement {
    let hCanvas = document.createElement('canvas')
    hCanvas.width = nCanvas.width
    hCanvas.height= nCanvas.height
    let hCtx = hCanvas.getContext('2d')
    let nCtx = nCanvas.getContext('2d')
    let nDat = nCtx.getImageData(0,0,nCanvas.width,nCanvas.height)
    hCtx?.putImageData(nDat, 0, 0)
    return hCanvas
}

export function toNodeCanvas(inCanvas : HTMLCanvasElement | SeqCanvas ) : Canvas {
    let nCanvas = null
    if( isSeqCanvas (inCanvas)  ) {
        // this is more inefficient, but allows to test more code
        let hCanvas =  document.createElement('canvas')
        CanvasUtils.fromSeq(inCanvas,hCanvas)
        nCanvas = toNodeCanvas(hCanvas)
    } else {
        nCanvas = createCanvas(inCanvas.width,inCanvas.height)
        let nCtx = nCanvas.getContext('2d')
        let hCtx = inCanvas.getContext('2d') as CanvasRenderingContext2D
        let hDat = hCtx.getImageData(0,0,nCanvas.width,nCanvas.height) 
        nCtx?.putImageData(hDat, 0, 0) 
    }
    return nCanvas
}

function isSeqCanvas(obj : any ) : obj is SeqCanvas {
    return obj.data != null && obj.width != null && obj.height != null && obj.data.length > 0 
}

async function saveCanvas( canvas : Canvas, filename : string ) : Promise<void>{
    return new Promise( (res,rej)=>{
        const out = fs.createWriteStream(`${filename}.png`)
        const stream = canvas.createPNGStream()
        stream.pipe(out)
        out.on('finish', () =>  res())
    })
}

const maxHeight = 512
export  function displayHisto(histo : Histogram , cMap? : string []) : Canvas {
    let maxVal= histo.maxValue
    let bins  = histo.histogramBins
    let scale = Math.min( 1, maxHeight/maxVal)
    let width = bins.length
    let height= Math.floor(scale*maxVal)
    let canvas = createCanvas(width,height)
    let ctx = canvas.getContext('2d')
    ctx.fillStyle="#FFFFFFFF"
    ctx.fillRect(0,0,width,height)
    bins.forEach((v,i)=>{
        let col =  cMap ? cMap[i] : "#0000FFFF"
        ctx.beginPath()
        ctx.strokeStyle= col
        let scaled = Math.floor(v * scale)
        ctx.moveTo(i,height)
        ctx.lineTo(i,height-scaled)
        ctx.stroke()
    })
    return canvas
}

const randomCoords =' \
0.182	0.941	0.731	0.569	0.690	0.887	0.988	0.130   \
0.665	0.971	0.741	0.356	0.132	0.871	0.199	0.852   \
0.393	0.612	0.458	0.677	0.426	0.649	0.814	0.373   \
0.088	0.010	0.415	0.751	0.443	0.397	0.114	0.819   \
0.300	0.507	0.321	0.202	0.081	0.114	0.656	0.116   \
0.655	0.552	0.956	0.870	0.602	0.876	0.297	0.394   \
0.831	0.566	0.352	0.252	0.356	0.297	0.972	0.810   \
0.954	0.567	0.747	0.723	0.190	0.377	0.634	0.655   \
0.931	0.617	0.787	0.653	0.491	0.509	0.751	0.276   \
0.474	0.509	0.082	0.354	0.334	0.315	0.600	0.204   \
0.413	0.280	0.968	0.293	0.247	0.138	0.471	0.230   \
0.246	0.304	0.398	0.871	0.548	0.626	0.963	0.870   \
0.289	0.193	0.042	0.250	0.617	0.441	0.248	0.332   \
0.988	0.192	0.079	0.742	0.387	0.047	0.848	0.266   \
0.979	0.289	0.349	0.271	0.452	0.182	0.614	0.361   \
0.909	0.262	0.522	0.650	0.810	0.561	0.340	0.873   \
0.622	0.441	0.154	0.277	0.044	0.247	0.002	0.769   \
0.230	0.449	0.075	0.155	0.189	0.224	0.463	0.439   \
0.583	0.834	0.544	0.728	0.853	0.866	0.160	0.762   \
0.362	0.661	0.310	0.225	0.009	0.024	0.125	0.443   \
0.003	0.834	0.899	0.645	0.552	0.914	0.985	0.627   \
0.889	0.236	0.633	0.534	0.710	0.270	0.355	0.654   \
0.554	0.631	0.507	0.675	0.252	0.523	0.685	0.112   \
0.668	0.934	0.304	0.334	0.426	0.322	0.486	0.861   \
0.215	0.286	0.680	0.632	0.348	0.359	0.807	0.131   \
0.583	0.622	0.664	0.878	0.552	0.704	0.928	0.528   \
0.001	0.912	0.560	0.657	0.547	0.655	0.057	0.961   \
0.373	0.535	0.322	0.108	0.290	0.313	0.951	0.939   \
0.163	0.243	0.874	0.724	0.401	0.894	0.203	0.074   \
0.687	0.443	0.736	0.987	0.984	0.047	0.465	0.488   \
0.605	0.581	0.194	0.778	0.514	0.221	0.096	0.622   \
0.664	0.553	0.350	0.868	0.690	0.141	0.263	0.435   \
0.377	0.611	0.680	0.095	0.228	0.614	0.997	0.587   \
0.149	0.452	0.598	0.396	0.802	0.871	0.450	0.143   \
0.459	0.312	0.525	0.360	0.440	0.291	0.378	0.277   \
0.539	0.152	0.078	0.890	0.845	0.319	0.278	0.545   \
0.500	0.822	0.785	0.052	0.109	0.243	0.753	0.090   \
0.853	0.479	0.345	0.408	0.968	0.503	0.943	0.052   \
0.434	0.363	0.487	0.367	0.479	0.090	0.281	0.032   \
0.983	0.981	0.227	0.562	0.381	0.816	0.492	0.046   \
0.063	0.390	0.640	0.683	0.298	0.108	0.448	0.359   \
0.232	0.360	0.755	0.615	0.230	0.933	0.462	0.079   \
0.090	0.096	0.188	0.801	0.653	0.007	0.499	0.408   \
0.849	0.722	0.376	0.007	0.023	0.099	0.365	0.859   \
0.220	0.974	0.740	0.871	0.904	0.137	0.176	0.010   \
0.514	0.418	0.316	0.837	0.812	0.901	0.109	0.365   \
0.059	0.838	0.715	0.713	0.695	0.447	0.472	0.824   \
0.966	0.626	0.627	0.004	0.484	0.982	0.467	0.768   \
0.207	0.136	0.007	0.864	0.891	0.260	0.579	0.711   \
0.222	0.614	0.675	0.816	0.910	0.375	0.726	0.952   \
0.876	0.111	0.097	0.757	0.808	0.200	0.815	0.059   \
0.584	0.954	0.431	0.643	0.678	0.810	0.452	0.272   \
0.374	0.649	0.218	0.183	0.042	0.489	0.041	0.660   \
0.237	0.243	0.917	0.807	0.357	0.626	0.634	0.104   \
0.800	0.295	0.267	0.704	0.111	0.078	0.156	0.519   \
0.541	0.288	0.983	0.200	0.362	0.293	0.862	0.669   \
0.605	0.348	0.742	0.607	0.786	0.393	0.168	0.011   \
0.542	0.088	0.007	0.512	0.529	0.143	0.044	0.976   \
0.686	0.991	0.930	0.025	0.873	0.494	0.383	0.671   \
0.595	0.537	0.092	0.415	0.561	0.083	0.596	0.916   \
0.477	0.048	0.566	0.999	0.885	0.639	0.131	0.963   \
0.749	0.621	0.469	0.631	0.787	0.462	0.584	0.316   \
0.442	0.690	0.800	0.099	0.822	0.580	0.350	0.936   \
0.316	0.766	0.889	0.701	0.948	0.368	0.837	0.725   \
0.254	0.185	0.222	0.407	0.203	0.034	0.288	0.128   \
0.429	0.893	0.456	0.439	0.150	0.521	0.937	0.087   \
0.578	0.398	0.266	0.978	0.200	0.153	0.392	0.829   \
0.828	0.418	0.224	0.508	0.792	0.040	0.750	0.154   \
0.852	0.597	0.771	0.553	0.206	0.319	0.993	0.745   \
0.488	0.691	0.666	0.376	0.077	0.782	0.665	0.297   \
0.311	0.560	0.534	0.090	0.547	0.423	0.989	0.206   \
0.478	0.726	0.216	0.581	0.207	0.280	0.313	0.414   \
0.717	0.346	0.101	0.831	0.483	0.362	0.118	0.454   \
0.258	0.419	0.180	0.476	0.346	0.370	0.827	0.506   \
0.426	0.555	0.030	0.026	0.340	0.536	0.062	0.372   \
0.241	0.185	0.901	0.837	0.608	0.027	0.417	0.306   \
0.503	0.002	0.247	0.440	0.160	0.833	0.910	0.709   \
0.631	0.501	0.163	0.369	0.931	0.970	0.331	0.871   \
0.814	0.939	0.707	0.060	0.148	0.410	0.476	0.706   \
0.963	0.374	0.794	0.520	0.288	0.800	0.482	0.136   \
0.441	0.909	0.254	0.753	0.298	0.694	0.122	0.067   \
0.177	0.757	0.921	0.537	0.165	0.153	0.674	0.262   \
0.090	0.986	0.387	0.699	0.834	0.243	0.524	0.081   \
0.577	0.468	0.059	0.889	0.977	0.076	0.503	0.451   \
0.372	0.767	0.806	0.663	0.398	0.272	0.065	0.532   \
0.604	0.355	0.141	0.054	0.464	0.162	0.819	0.693   \
0.580	0.339	0.118	0.261	0.265	0.381	0.313	0.755   \
0.463	0.876	0.069	0.831	0.463	0.958	0.329	0.209   \
0.961	0.618	0.510	0.987	0.403	0.813	0.503	0.735   \
0.646	0.540	0.166	0.058	0.269	0.633	0.393	0.448   \
0.555	0.772	0.774	0.107	0.483	0.442	0.716	0.130   \
0.317	0.616	0.500	0.335	0.822	0.697	0.381	0.290   \
0.238	0.514	0.396	0.147	0.347	0.588	0.642	0.319   \
0.718	0.092	0.296	0.188	0.232	0.528	0.353	0.980   \
0.357	0.872	0.159	0.737	0.180	0.799	0.696	0.831   \
0.750	0.381	0.483	0.659	0.674	0.322	0.130	0.554   \
0.648	0.435	0.721	0.519	0.295	0.860	0.296	0.952   \
0.035	0.035	0.994	0.023	0.454	0.635	0.850	0.442   \
0.144	0.017	0.732	0.414	0.810	0.848	0.412	0.303   \
0.749	0.463	0.725	0.292	0.428	0.269	0.761	0.668   \
'

/**
 * return a list of number from 0 to 1 distribuited randomly.
 * it provies the same list at each run for consistency in the
 * tests
 */
export function randomNumber() : number [] {
    let vals = randomCoords.split(/[\t\s]+/)
    let numbers : number [] = []
    vals.forEach((v)=> v.trim() != '' ? numbers.push(parseFloat(v)) : null )
    return numbers
}

/**
 * generate a lits of random points on the width, height range.
 * The pointts are generte in interger coordinates.
 * @param width range of the x coordinates
 * @param height range of the y coordinates
 */
export function randomPoints( width : number, height : number ) : {x:number, y:number } [] {
    let vals = randomNumber();
    let points : {x:number, y:number } [] = []
    for( let i=0; i<Math.floor(vals.length/2); i++ ){
        let x = Math.floor( vals[2*i] * width )
        let y = Math.floor( vals[2*i+1] * height )
        points.push({x,y})
    }
    return points
}

export function overlayTensor( canvas : Canvas, tf : TensorField, size : number =2 ) : Canvas {
    let mag = 2*size+1
    let width = canvas.width
    let height= canvas.height
    let outCanvas = createCanvas(width*mag, height*mag)

    // normalize the tensor field
    let maxEgVal = tf.fieldTensors.reduce((m,v)=>Math.max(m,v.majVal),0)
    let rThr = maxEgVal * 2/3 // red threshold : top 30%
    let gThr = maxEgVal * 1/3 // greeen threshold : mid 30%
      // blue threshold is bottom 30%

    // drawing
    let ctx = outCanvas.getContext('2d')
    ctx.drawImage(canvas, 0, 0, width, height, 0, 0, outCanvas.width, outCanvas.height )

    let step = 2
    let mjLen = step*size

    for( let y=0; y<height; y+=step){
        for( let x=0; x<width; x+=step){
            let t = tf.get(x,y)
            let xPos = x*mag+size+1
            let yPos = y*mag+size+1
            let col = "#0000FF"
            if( t.majVal > gThr ) col = "#00FF00"
            if( t.majVal > rThr ) col = "#FF0000"
            let xMjS = xPos+t.majVec[0] * mjLen
            let yMjS = yPos+t.majVec[1] * mjLen
            let xMjE = xPos-t.majVec[0] * mjLen
            let yMjE = yPos-t.majVec[1] * mjLen

            ctx.beginPath()
            ctx.strokeStyle = col
            ctx.moveTo(xMjS,yMjS)
            ctx.lineTo(xMjE,yMjE)
            if( t.minVal > 0 ){
                let mnlen = Math.max( mjLen * t.minVal/t.majVal, 1 )
                let xMnS = xPos+t.minVec[0] * mnlen
                let yMnS = yPos+t.minVec[1] * mnlen
                let xMnE = xPos-t.minVec[0] * mnlen
                let yMnE = yPos-t.minVec[1] * mnlen     
                ctx.moveTo(xMnS,yMnS)
                ctx.lineTo(xMnE,yMnE)  
            }
            ctx.stroke()
        }
    }
    return outCanvas
}


export function overlayTangent( canvas : Canvas, tf : TensorField, size : number =2 ) : Canvas {
    let mag = 2*size+1
    let width = canvas.width
    let height= canvas.height
    let outCanvas = createCanvas(width*mag, height*mag)

    // normalize the tensor field
    let maxEgVal = tf.fieldTensors.reduce((m,v)=>Math.max(m,v.minVal),0)
    let rThr = maxEgVal * 2/3 // red threshold : top 30%
    let gThr = maxEgVal * 1/3 // greeen threshold : mid 30%
      // blue threshold is bottom 30%

    // drawing
    let ctx = outCanvas.getContext('2d')
    ctx.drawImage(canvas, 0, 0, width, height, 0, 0, outCanvas.width, outCanvas.height )

    let step = 2
    let mjLen = step*size

    for( let y=0; y<height; y+=step){
        for( let x=0; x<width; x+=step){
            let t = tf.get(x,y)
            let xPos = x*mag+size+1
            let yPos = y*mag+size+1
            let col = "#0000FF"
            if( t.majVal > gThr ) col = "#00FF00"
            if( t.majVal > rThr ) col = "#FF0000"
            if( t.minVal > 0 ){
                ctx.beginPath()
                ctx.strokeStyle = col
                let xMnS = xPos+t.minVec[0] * mjLen
                let yMnS = yPos+t.minVec[1] * mjLen
                let xMnE = xPos-t.minVec[0] * mjLen
                let yMnE = yPos-t.minVec[1] * mjLen     
                ctx.moveTo(xMnS,yMnS)
                ctx.lineTo(xMnE,yMnE)  
                ctx.stroke()
            }
        }
    }
    return outCanvas
}

export default {}