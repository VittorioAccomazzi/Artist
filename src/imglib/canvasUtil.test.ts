import {hash, getCanvases, toHTMLCanvas, toNodeCanvas} from './testutils'
import CanvasUtils from './canvasUtils'

test('shall split and combine canvas in images',async ()=>{
    for await ( const [canvas1,] of getCanvases() ) {
        let hsh1 = await hash(canvas1)
        const [ r, g, b, a ] = CanvasUtils.Split( toHTMLCanvas(canvas1))
        let canvas2 = toNodeCanvas(CanvasUtils.Compose(r,g,b,a))
        let hsh2 = await hash(canvas2)
        expect(hsh1).toBe(hsh2)
        let canvas3 = toNodeCanvas(CanvasUtils.Compose(r,g,b))
        let hsh3 = await hash(canvas3)
        expect(hsh1).toBe(hsh3)
    }
})

test('shall convert to gray scale',async ()=>{
    for await ( const [canvas,] of getCanvases() ) {
        let int = CanvasUtils.Intensity(toHTMLCanvas(canvas))
        let hsh = await hash(int)
        expect(hsh).toMatchSnapshot()
    }
})