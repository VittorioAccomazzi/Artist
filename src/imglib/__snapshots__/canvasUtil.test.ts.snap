// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall convert to gray scale 1`] = `"0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007dd000000000000000000000000000f7fffff0000000000000000000000009ffffffdf00000000000000000000003fffffffffd000000000000000000002fffffffffff80000000000000000002fffffffffffff8000000000000000007ffffffffffffffd0000000000000007ffffffffffffffffe00000000000000ffffffffffffffffffe000000000000ffffffffffffffffffffc08000000007ffffffffffffffffffffff800000011fffffffffffffffffffffffe8000002bffffffffffffffffffffffffde4000fffffffffffffffffffffffffffff00bfffffffffffffffffffffffffffff00fffffffffffffffffffffffffffffff0fffffffffffffffffffffffffffffff8fffffffffffffffffffffffffffffffeffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00000000000000000000000000000000000000000000000000000160000000000000000000000830000000c0000000000000000000003ffbd00000000000000000000000001ffffffe1f00000000000000000000007fffffffcc0800000000000000000003fffffffffe100000000000000000000fffffffffffa00000000000000000007fffffffffff80000000000000000000ffffffffffff80000000000000000000ffffffffffffc0000000000000000003fffffffffffff000000000000000000ffffffffffffffe00000000000000003ffffffffffffffe0000000000000001ffffffffffffffffc000000000000003fffffffffffffffff800000000000007ffffffffffffffffffe000000000001ffffffffffffffffffff400000000007fffffffffffffffffffff0000000001ffffffffffffffffffffff0000000003fffffffffffffffffffffff80000000ffffffffffffffffffffffffe0000003ffffffffffffffffffffffffe800000ffffffffffffffffffffffffffc00001fffffffffffffffffffffffffffe001fffffffffffffffffffffffffffff00fffffffffffffffffffff7fffd002000fffffffffffffffffffe620130000000fffffffffffffffffffa000000000000ffffffffffffffffff80000000000000fffffffffffffc000000000000000000fffffffffff8d0000000000000000000fffffffffffffbffffe0000000000000fffffffffffd80fbf4f0000006000000fffffffffffc005e7068000000000000ffffffffffc004e40000000000000000fffff0000000000000000000000000000000000000000000000000000000000000000000000000037ffffff80000000000000000000003bfffffffffffffffff0000000000effffffffffffffffffffb000000003ffffffffffffffffffffffb00000007ffffffffffffffffffffff7f00000007fffffffffffffffffc53ffff0000000ffffffffffffffffffffffffb00160ffffffffffffffffffffefe3ff67013fffffffffffffffffffffc008001837ffffffffffffffffffdf800000000fffffffffffffffffffffffd00000000fffffffffffffffffffffe0000000000ffffffffffffffffffffc20000000000fffffffffffffffffffffff380000000ffffffffffffffffffffe00000000000ffffffffe7ff9fffffffff4000000000fffffffffffbfffe03ffa000000000007fcffffffc3ff81ffffc000000000000dfffff00fff7ffff8040000000000000e7cb7f7ffc7ffd7000000000000000001003ffc5ffff960e3f820000000000000000634fff80007df000000000000000808006780004f24000000000000000000003ff800ffffb80003f00000000000000000002fff00000020000000000000003801c07f80000018000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffbffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8ffffffffffffffffffffffffffffffeffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff4441cfffffffbfffffffffffffffff0fffffffffffffcbffffffffffff801fffffffe077ffffffffffffffffffffffff000700087ffffffffff7fffffffeffffffffd9e7f8ff3e7e3dffbffffffffdf800003ffffffeffb1fffffffff7ff800000000ffff101fffffff87381e07941ffffe0c000065fdfe7ffffffdfff38007ff0183c000007ff02201f8f3f0040c0000007dfc00010007c7fffffffe0000100000000000000000180000fff89ffff5f8039e00000000d0000000ffff400000000000000000018fb8031f0000000000007800000018c07ffffff000000000000180000000400071fff8074000000061fc0000000000fd02004000a300004e9e00000000000000000000001ef00001e00000000000000000000000ff0000000000000000000000000000008000000000000000000000000000000000078300000000000000000000000000fff8000300000000000000000000000f000000080000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall convert to gray scale 2`] = `"fffe0000b5ffffffbd007fffffc00003fffe0000fdffffffd0003fffffc00000fffe00003ffffcdfe0003fffffe00010fffe00001fffffdfe1001fffffe00010fffe00005fffffdfecd01ffffff000e0fffe00007ffffbcfe0001ffffff001c0fffe000015dbc007c0001ffffff80180fffe000003f9c003c0001ffffffc0100fffe000001bdc002c0000ffffffc0200fffe0000000c8001c0000ffffffe0000fffe00000000800084000ffffffe0000fffe000000388ff800000fffffff0000fffe00000018ffff06000fffffff8000fffe00000001ffffc0000fffffff8000fdfe0000003bffffe0000fffffffc000f9fe00000171fffff8000fffffff8000f1fe000001037ffffc000fffffff8000f1fe000000017ffffe000fffffff0000e1fe00000200ffffff800fe7fffe0000e1fe000004017fffffc00fe3fffc0001c1fe000000007fffffe00fe3fff8000381fe00000000fffffff00fe1fff0000381fe00000000fffffff80fe0fff0000701fe00000001fffffff80fe0ffe0000f01fe00000001fffffffe0fe07fc0001f01fe0000400bffffffff0fe07fc0003f01fe0000400ffeffffff0fe03f80007f01fe00008007f1ffffff8fe01f0000ff01fe0000800fc7ffffffcfe03f0000ff01fe0001801f8fffffffcfe07f0001ff01fe000180231fffffffefe0ff0003ff01fe000180443fffffffefe3ff8007ff00fe000180007ffffffff7c7ff8007ff00fe00030000ffffffffffdfff800fff00fe00030001ffffffffffffff000fff00fe00030007ffffffffffffff001fff00fe00038007bfffffffffffff001fff00fe0003800f7fffffffffffff003fff00fe0003c01efffffffffffffe007fff00fe0003e029fffffffffffffe007fff00fe0003e043fffffffffffffe00ffff00fe0003e087bffffffffffffc00ffff00fe0003f00a3ff1ffffffffcc01ffff00fe0003f8085c0c7c7fffff0c01ffff01fe0003f8101a0c007ffffe1803ffff01fe0003c0002000007ffff03003ffff01fe0003c800400001ffffc0e007ffff01fe0003d800800002ffff03c007ffff01fe0001f800400005fffc3f800fffff00fe0001f00000000bfff83f000fffff00fe0001f00000000ffff03e000fffff00fe0000f00000000ffff03c0007ffff00fe0000f00000001ffff0300007ffff00fe0000700000003ffff8300007ffff00fe0000700000007ffff830000fffff00fe000060000001fffff830001fffff00fe000060000003fffffc10005fffff00fe000060000007fffffc10007fffff00fe000020000007fffffc1000bfffff00fe00000000000ffffffc1000bfffff00fe00001000001ffffffc10007fffff00fe00018800003fc1fffc10007ffffe00fe00010000007fb07ff008007ffffc00fe0003800000fe083fc00800fffffc01fe000fc00000fc003fc00803ffffff00fe001f800003f800bf000803ffffff00fe001e000003f00c3f080803ffffff00fe001e000003f00e1e080807ffffff00fe001c000007e63e3e380807ffffff00ff00700000079f977fc00807ffffff00ff01c000001f1fffffe0080fffffff00ff010000000f3ffffff0080fffffff00ff038000403e1fffeff8080fffffff00ff070000001c3ffffff80c1fffffff00ff00000010781ffffff80c1fffffff00ff00000008383fffdff80c1fffffff00ff04000000703fffcff80c1fffffff00ff08000030603fffcff80c3fffffff00ff00010000c03fff4ff8043fffffff00ff00000003c01ffef7f0063fffffff00ff00000003801ffe87f0067fffffff00ff00008007801fff03f0067fffffff01ff0000040b001fffffe0067fffffff01ff04000032001fffffe0067fffffff01ff08002070001fffffe002ffffffff01ff00000060001fffffc002ffffffff01ff000011c8001fffffc002ffffffff01ff080007d00017c0008003ffff7fff01ff080407900007e0ff8003ffffffff01ff90040f600003f1ff0003ffffffff01ff00000fc00003f8fe0003ffff7fff00ff200823800001ff1e0003fffe7fff00ff000903000001fffc0003fffe7fff01ff000003000001fffc0003fffe7fff01ff000098000000fffc0003fffeffff01ff0000280000003ff80003fffefffe01ff20041f0000001ff80003fffffff801ffc0042d800003fffe00031fffffe000ff800664c00007ffff8003e0ffffc000ff200041800007ffffe003f800ff80c0ff400147a00003fffff803ff04ff00e0ffc0000f100007fffffc03ff84fe00f8ff800007100007fffffe03ffedfc0078ffc00003a00007ffffff07fffffc003cff800003d80043ffffff07fffffc003cff800001c80047ffffff87fffff8003cff800001880007ffffffc7fffff8003cff800001c80007ffffffc7fffff0003cff8040019800c7ffffffc7fffff0001cff000009f000c7ffffffe7fffff0001cff8000033801c7ffffffe7fffff0001cff8000038001c7ffffffe7fffff0001cff8000014001cffffffff7ffffe0001cff8000072001cfffffffffffffc0001cff80000130038ffffffffdffffc0001cff80000000039ffffffff81fffc0001e7f80000800071ffffffff801ff80001e7f00000400063ffffffff8007f80001e7f00000600063ffffffff8007f00001e7f800003000c7ffffffffc007e00001e7f8000008010fffffffffc007c00003e7f8000000000fffffffffc007800001e7f0000010003fffffffffe007000000e7f000000000ffffffffffe006000400f7f000000001ffffffffffe070000800ffe000000003ffffffffffe0f8801800f7f000080003fffffffffff07b803000ffe000000007fffffffffff07f80300"`;
